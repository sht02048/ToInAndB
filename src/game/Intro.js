import Canvas from "./Canvas";
import Sprite from "./Sprite";

class Intro extends Canvas {
  #floatSpeed = 0.1;

  constructor(titlePath, instructionPath) {
    super("intro-canvas");

    this.x = this.canvas.width / 2;
    this.title = new Sprite(titlePath);
    this.instruction = new Sprite(instructionPath);
    this.isUp = true;
    this.amplitude = 0;
  }

  float() {
    if (this.isUp) {
      this.y += this.#floatSpeed;
    } else {
      this.y -= this.#floatSpeed;
    }

    this.amplitude += 1;

    if (this.amplitude === 100) {
      this.amplitude = 0;
      this.isUp = !this.isUp;
    }

    this.ctx.drawImage(
      this.title,
      this.x - this.title.width / 4,
      100,
      this.title.width / 2,
      this.title.height / 2,
    );

    this.ctx.drawImage(
      this.instruction,
      this.x - this.instruction.width / 4,
      this.y + this.canvas.height - 150,
      this.instruction.width / 2,
      this.instruction.height / 2,
    );
  }

  out() {
    this.y += this.inAndOutSpeed;

    this.ctx.drawImage(
      this.title,
      this.x - this.title.width / 4,
      this.y + 100,
      this.title.width / 2,
      this.title.height / 2,
    );

    this.ctx.drawImage(
      this.instruction,
      this.x - this.instruction.width / 4,
      this.y + this.canvas.height - 150,
      this.instruction.width / 2,
      this.instruction.height / 2,
    );
  }
}

export default Intro;
