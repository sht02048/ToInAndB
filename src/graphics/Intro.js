import Sound from "../utils/Sound";
import { IMAGE, AUDIO } from "../constants/path";
import Renderer from "./Renderer";

class Intro {
  #floatSpeed = 0.1;
  #titleWidth = 1158;
  #titleHeight = 951;
  #instructionWidth = 934;
  #instructionHeight = 53;

  constructor() {
    this.title = new Renderer(IMAGE.TITLE);
    this.instruction = new Renderer(IMAGE.INSTRUCTION_START);

    this.canvasWidth = this.title.canvasWidth;
    this.canvasHeight = this.title.canvasHeight;
    this.inAndOutSpeed = this.title.inAndOutSpeed;

    this.x = this.canvasWidth / 2;
    this.instructionY = this.canvasHeight - 150;
    this.titleY = 100;

    this.isUp = true;
    this.amplitude = 0;
    this.shouldBeDisplayed = false;

    this.introMusic = new Sound(AUDIO.INTRO);
    this.battleMusic = new Sound(AUDIO.BATTLE);
  }

  float() {
    if (this.isUp) {
      this.instructionY += this.#floatSpeed;
    } else {
      this.instructionY -= this.#floatSpeed;
    }

    this.amplitude += 1;

    if (this.amplitude === 100) {
      this.amplitude = 0;
      this.isUp = !this.isUp;
    }
  }

  out() {
    this.titleY += this.inAndOutSpeed;
    this.instructionY += this.inAndOutSpeed;
  }

  render() {
    this.title.renderIntro(
      this.x - this.#titleWidth / 4,
      this.titleY,
      this.#titleWidth / 2,
      this.#titleHeight / 2,
    );
    this.instruction.renderIntro(
      this.x - this.#instructionWidth / 4,
      this.instructionY,
      this.#instructionWidth / 2,
      this.#instructionHeight / 2,
    );
  }

  playIntroMusic() {
    const muteToggle = document.getElementById("mute-toggle");

    muteToggle.addEventListener("mousedown", (event) => {
      event.preventDefault();

      if (!Sound.hasIntroPlayed) {
        this.introMusic.playAudio();
        Sound.hasIntroPlayed = true;
      }

      if (Sound.isPlaying) {
        Sound.mute();
      } else {
        Sound.unmute();
      }
    });
  }

  playBattleMusic() {
    this.introMusic.pauseAudio();
    this.battleMusic.playAudio();
  }
}

export default Intro;
