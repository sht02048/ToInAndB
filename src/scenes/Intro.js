import Sound from "../utils/Sound";
import Sprite from "../entities/Sprite";
import { IMAGE, AUDIO } from "../constants/path";

class Intro {
  #floatSpeed = 0.1;

  constructor(game) {
    this.game = game;

    this.title = new Sprite(IMAGE.TITLE);
    this.instruction = new Sprite(IMAGE.INSTRUCTION_START);

    this.x = this.game.introCanvas.width / 2;
    this.instructionY = this.game.introCanvas.height - 150;
    this.titleY = 100;

    this.isUp = true;
    this.amplitude = 0;

    this.introMusic = new Sound(AUDIO.INTRO);
    this.battleMusic = new Sound(AUDIO.BATTLE);
  }

  float() {
    if (this.isUp) {
      this.instructionY += this.#floatSpeed;
    } else {
      this.instructionY -= this.#floatSpeed;
    }

    this.amplitude += 1;

    if (this.amplitude === 100) {
      this.amplitude = 0;
      this.isUp = !this.isUp;
    }
  }

  out() {
    this.titleY += this.game.inAndOutSpeed;
    this.instructionY += this.game.inAndOutSpeed;
  }

  render() {
    this.game.introCtx.drawImage(
      this.title,
      this.x - this.title.width / 4,
      this.titleY,
      this.title.width / 2,
      this.title.height / 2,
    );

    this.game.introCtx.drawImage(
      this.instruction,
      this.x - this.instruction.width / 4,
      this.instructionY,
      this.instruction.width / 2,
      this.instruction.height / 2,
    );
  }

  playIntroMusic() {
    const muteToggle = document.getElementById("mute-toggle");

    muteToggle.addEventListener("mousedown", (event) => {
      event.preventDefault();

      this.introMusic.playAudio();
      Sound.toggleSound(muteToggle);
    });
  }

  playBattleMusic() {
    this.introMusic.pauseAudio();
    this.battleMusic.playAudio();
  }
}

export default Intro;
