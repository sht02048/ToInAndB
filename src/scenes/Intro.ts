import Sound from "../utils/Sound";
import { IMAGE, AUDIO } from "../constants/path";
import Renderer from "../graphics/Renderer";

class Intro {
  #floatSpeed = 0.1;
  #titleWidth = 1158;
  #titleHeight = 951;
  #instructionWidth = 934;
  #instructionHeight = 53;

  private title: Renderer;
  private instruction: Renderer;
  private introMusic: Sound;
  private battleMusic: Sound;
  private canvasWidth: number;
  private canvasHeight: number;
  private inAndOutSpeed: number;
  private x: number;
  private instructionY: number;
  private titleY: number;
  private isUp: boolean;
  private amplitude: number;
  private shouldBeDisplayed: boolean;

  constructor() {
    this.title = new Renderer(IMAGE.TITLE);
    this.instruction = new Renderer(IMAGE.INSTRUCTION_START);

    this.canvasWidth = this.title.canvasWidth;
    this.canvasHeight = this.title.canvasHeight;
    this.inAndOutSpeed = this.title.inAndOutSpeed;

    this.x = this.canvasWidth / 2;
    this.instructionY = this.canvasHeight - 150;
    this.titleY = 100;

    this.isUp = true;
    this.amplitude = 0;
    this.shouldBeDisplayed = false;

    this.introMusic = new Sound(AUDIO.INTRO);
    this.battleMusic = new Sound(AUDIO.BATTLE);
  }

  float() {
    if (this.isUp) {
      this.instructionY += this.#floatSpeed;
    } else {
      this.instructionY -= this.#floatSpeed;
    }

    this.amplitude += 1;

    if (this.amplitude === 100) {
      this.amplitude = 0;
      this.isUp = !this.isUp;
    }
  }

  out() {
    this.titleY += this.inAndOutSpeed;
    this.instructionY += this.inAndOutSpeed;
  }

  render() {
    this.title.renderIntro(
      this.x - this.#titleWidth / 4,
      this.titleY,
      this.#titleWidth / 2,
      this.#titleHeight / 2,
    );
    this.instruction.renderIntro(
      this.x - this.#instructionWidth / 4,
      this.instructionY,
      this.#instructionWidth / 2,
      this.#instructionHeight / 2,
    );
  }

  playIntroMusic() {
    const muteToggle = document.getElementById("mute-toggle");

    muteToggle.addEventListener("mousedown", (event) => {
      event.preventDefault();

      if (!Sound.hasIntroPlayed) {
        this.introMusic.playAudio();
        Sound.hasIntroPlayed = true;
      }

      if (Sound.isPlaying) {
        Sound.mute();
      } else {
        Sound.unmute();
      }
    });
  }

  playBattleMusic() {
    this.introMusic.pauseAudio();
    this.battleMusic.playAudio();
  }
}

export default Intro;
