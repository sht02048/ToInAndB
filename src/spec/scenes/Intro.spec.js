import Intro from "../../scenes/Intro";

describe("Intro", () => {
  const mainCanvas = document.createElement("canvas");
  const introCanvas = document.createElement("canvas");
  mainCanvas.id = "main-canvas";
  introCanvas.id = "intro-canvas";

  document.body.appendChild(mainCanvas);
  document.body.appendChild(introCanvas);

  let intro;

  beforeEach(() => {
    intro = new Intro();
  });

  it("should have proper functions", () => {
    expect(intro.float).to.be.a("function");
    expect(intro.out).to.be.a("function");
    expect(intro.render).to.be.a("function");
    expect(intro.playIntroMusic).to.be.a("function");
    expect(intro.playBattleMusic).to.be.a("function");
  });

  it("should out", () => {
    const originalTitleY = intro.titleY;
    const originalInstructionY = intro.instructionY;
    const inAndOutSpeed = 4;

    intro.out();

    expect(originalTitleY).to.equal(intro.titleY - inAndOutSpeed);
    expect(originalInstructionY).to.equal(intro.instructionY - inAndOutSpeed);
  });

  it("should render title and instruction", () => {
    intro.render();

    const ctx = introCanvas.getContext("2d");
    const events = ctx.__getEvents();

    expect(events).toMatchSnapshot();
  });

  describe("should float", () => {
    const flowSpeed = 0.1;
    let originalInstructionY;
    let originalIsUp;

    beforeEach(() => {
      originalInstructionY = intro.instructionY;
      originalIsUp = intro.originalIsUp;
    });

    it("go up when isUp is true", () => {
      intro.float();

      expect(originalInstructionY).to.equal(intro.instructionY - flowSpeed);
    });

    it("go down when isUp is false", () => {
      intro.isUp = false;
      intro.float();

      expect(originalInstructionY).to.equal(intro.instructionY + flowSpeed);
    });

    it("change the direction when amplitude is 100", () => {
      intro.amplitude = 99;

      intro.float();

      expect(originalIsUp).to.not.equal(intro.isUp);
    });
  });
});
