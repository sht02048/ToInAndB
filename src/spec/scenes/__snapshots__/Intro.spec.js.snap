// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Intro > should render title and instruction 1`] = `
[
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -139.5,
      "dy": 100,
      "img": <img
        src="/images/title.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  {
    "props": {
      "dHeight": 0,
      "dWidth": 0,
      "dx": -83.5,
      "dy": 0,
      "img": <img
        src="/images/instruction_start.png"
      />,
      "sHeight": 0,
      "sWidth": 0,
      "sx": 0,
      "sy": 0,
    },
    "transform": [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
